File: Add Binary.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *	 int val;
 *	 ListNode *next;
 *	 ListNode(int x) : val(x), next(NULL) {}
 * };
 */

#define F(i, a, b) for( int i = a; i <= b; i++ )
#define FB(i, a, b) for( int i = a; i >= b; i-- )

class Solution {
public:
	string addBinary(string a, string b) {
		int m = a.size();
		int n = b.size();
		if( m == 0 ){
			return b;
		}
		if( n == 0 ){
			return a;
		}

		if( m > n ){ 
			//m larger 
			int len = m; 
			int carry = 0; 
			int diff = m - n;
			FB( i, len-1, 0 ){ 
				int sum = carry;
				sum += a[i]-'0';
				if( i-diff > -1 ){ sum += b[i-diff]-'0'; }

				if( sum > 1 ){
					carry = 1;
					sum -= 2;
				}else{
					carry = 0;
				}

				if( sum == 0 ){
					a[i] = '0';
				}else{
					a[i] = '1';
				}
			}
			if( carry == 1 ){
				a.insert( a.begin(), '1' ); 
			}
			return a;
		}else{ 
			//n larger 
			int len = n;
			int carry = 0;
			int diff = n - m;
			FB( i, len-1, 0 ){
				int sum = carry;
				if( i-diff > -1 ){ sum += a[i-diff]-'0'; }
				sum += b[i]-'0';

				if( sum > 1 ){
					carry = 1;
					sum -= 2;
				}else{
					carry = 0;
				}

				if( sum == 0 ){
					b[i] = '0';
				}else{
					b[i] = '1';
				}
			}
			if( carry == 1 ){
				b.insert( b.begin(), '1' );
			}
			return b;
		}
	}
};

File: LeetCode-Java/Add Binary.java
public class Solution {
	public static String addBinary(String a, String b) {
		if( a.length() < b.length() ){
			String tmp = a;
			a = b;
			b = tmp;
		}
		char [] res = new char [a.length()+1];
		String ret;
		int m = a.length()-1;
		int n = b.length()-1; 
		int carry = 0;

		while( m >= 0 ){
			if( n < 0 ){
				int sum = a.charAt(m)-'0' + carry;
				res[m+1] = (char)(sum%2 + 48);
				carry = sum/2;
			}else{
				int sum = a.charAt(m)-'0' + b.charAt(n)-'0' + carry;
				res[m+1] = (char)(sum%2 + 48);
				carry = sum/2;
			}
			m--;
			n--;
		}
		
		res[0] = (char)(carry + 48);
		ret = new String(res);
		if( res[0] == '0' ){
			ret = ret.substring(1, a.length()+1);
		}
			
		return ret; 
	}
}
